File: cpp/structural/proxy.cpp
/*!
* \brief. Proxy. Provide a surrogate or placeholder for
*         another object to control access to it.
*
*                 Subject
*                 |     |    
*       RealSubject <-  Proxy
*/
#include <iostream>

class Subject {
public:
  virtual void Request() = 0;
};

class RealSubject : public Subject {
public:
  void Request() {
    std::cout << "RealSubject Request" << std::endl;
  }
};

class Proxy : public Subject {
public:
  Proxy(): sub_(NULL) {}
  void SetSubject(Subject* sub) {
    sub_ = sub;
  }

  void Request() {
    if (sub_ != NULL)
      sub_->Request();
    else
      std::cout << "No Subject." << std::endl;
  }

private:
  Subject* sub_;
};

int main() {
  Subject* sub = new RealSubject();
  Proxy* p = new Proxy();

  p->SetSubject(sub);
  // Use a proxy to do something for the subject.
  p->Request();

  delete sub;
  delete p;

  return 0;
}

File: python/structural/proxy.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

""" 
brief:
  Proxy. 
    Provide a surrogate or placeholder for
  another object to control access to it.

              Subject
              |     |    
    RealSubject <-  Proxy
"""

class RealSubject(object):
  def request(self):
    print("RealSubject Request")
    
class Proxy(object):
  def __init__(self):
    self._subject = None
    
  def set_subject(self, subject):
    self._subject = subject
    
  def request(self):
    if self._subject != None:
      self._subject.request()
    else:
      print("No Subject.")

if __name__ == '__main__':

  subject = RealSubject()
  proxy = Proxy()
  proxy.set_subject(subject)
  # Use a proxy to do something for the subject.
  proxy.request()